### PR TITLE
Make api consistent

### PR DESCRIPTION
**Pull request by ahartel**

### CODE REVIEW

File: aleph_alpha_client/completion.py [hunk 1]
```diff
@@ -0,0 +1,116 @@
...
+from aleph_alpha_client.prompt import _to_serializable_prompt
+
+
+class CompletionRequest(NamedTuple):
+    """
+    Describes a completion request
+
+    Parameters:
+        prompt (str, optional, default ""):
+            The text to be completed. Unconditional completion can be started with an empty string (default). The prompt may contain a zero shot or few shot task.
```
Comment (volkerstampa) [inline aleph_alpha_client/completion.py:13]: Image prompt?

File: aleph_alpha_client/completion.py [hunk 2]
```diff
@@ -0,0 +1,116 @@
...
+
+        maximum_tokens (int, optional, default 64):
+            The maximum number of tokens to be generated. Completion will terminate after the maximum number of tokens is reached. Increase this value to generate longer texts. A text is split into tokens. Usually there are more tokens than words. The summed number of tokens of prompt and maximum_tokens depends on the model (for luminous-base, it may not exceed 2048 tokens).
```
Comment (volkerstampa) [inline aleph_alpha_client/completion.py:16]: The _maximal supported_ (?) summed number of tokens of prompt and maximum_tokens depends ...

File: aleph_alpha_client/completion.py [hunk 3]
```diff
@@ -0,0 +1,116 @@
...
+
+        top_p (float, optional, default 0.0)
+            Introduces random sampling for generated tokens by randomly selecting the next token from the smallest possible set of tokens whose cumulative probability exceeds the probability top_p. Set to 0.0 if repeatable output is to be produced.
+            It is recommended to use either temperature, top_k or top_p and not all at the same time. If a combination of temperature, top_k or top_p is used rescaling of logits with temperature will be performed first. Then top_k is applied. Top_p follows last.
+
+        presence_penalty (float, optional, default 0.0)
+            The presence penalty reduces the likelihood of generating tokens that are already present in the text. Presence penalty is independent of the number of occurences. Increase the value to produce text that is not repeating the input.
+
+        frequency_penalty (float, optional, default 0.0)
+            The frequency penalty reduces the likelihood of generating tokens that are already present in the text. Presence penalty is dependent on the number of occurences of a token.
```
Comment (volkerstampa) [inline aleph_alpha_client/completion.py:35]: **Presence -> Frequency** penalty is dependent on the number of occurences

File: aleph_alpha_client/completion.py [hunk 4]
```diff
@@ -0,0 +1,116 @@
...
+            return tokens of completion
+
+        disable_optimizations  (bool, optional, default False)
+            We continually research optimal ways to work with our models. By default, we apply these optimizations to both your prompt and  completion for you.
+
+            Our goal is to improve your results while using our API. But you can always pass disable_optimizations: true and we will leave your prompt and completion untouched.
+    """
+
+    prompt: Union[str, List[Union[str, ImagePrompt]]] = ""
+    maximum_tokens: Optional[int] = 64
```
Comment (volkerstampa) [inline aleph_alpha_client/completion.py:71]: Does it make sense to have those parameters with not-`None` default `Optional`?

File: aleph_alpha_client/detokenization.py [hunk 5]
```diff
@@ -0,0 +1,19 @@
+from typing import Any, Dict, List, NamedTuple, Optional, Sequence
+
+
+class DetokenizationRequest(NamedTuple):
+    token_ids: List[int]
```
Comment (volkerstampa) [inline aleph_alpha_client/detokenization.py:5]: `Sequence`?

File: aleph_alpha_client/completion.py [hunk 6]
```diff
@@ -0,0 +1,116 @@
...
+            "prompt": _to_serializable_prompt(self.prompt),
+            "maximum_tokens": self.maximum_tokens,
+            "temperature": self.temperature,
+            "top_k": self.top_k,
+            "top_p": self.top_p,
+            "presence_penalty": self.presence_penalty,
+            "frequency_penalty": self.frequency_penalty,
+            "best_of": self.best_of,
+            "n": self.n,
+            "logit_bias": self.logit_bias,
```
Comment (volkerstampa) [inline aleph_alpha_client/completion.py:100]: ? A dict int -> float does not render as valid json, does it?

File: aleph_alpha_client/completion.py (hunk 2 above)
Comment (ahartel) [reply aleph_alpha_client/completion.py:16]: This is not the maximal supported number. The maximal supported number of tokens is fixed and defined by the model.
Here, the user determines how many tokens he wants to have generated at most.
So it's more like the _maximal desired_ number of tokens.

File: aleph_alpha_client/completion.py (hunk 4 above)
Comment (ahartel) [reply aleph_alpha_client/completion.py:71]: nope

File: aleph_alpha_client/detokenization.py (hunk 5 above)
Comment (ahartel) [reply aleph_alpha_client/detokenization.py:5]: done

File: README.md [hunk 7]
```diff
@@ -614,3 +317,14 @@ If an html coverage report has been created a simple http server can be run to s
 ```bash
 python -m http.server --directory htmlcov 8000
 ```
+
+## Update README
+
+> Do not change the README.md directly as it is generated from readme.ipynb 
+
```
[the remark this answers is not in the document]
Comment (ahartel) [reply README.md:324]: Good idea

File: tests/test_detokenize.py [hunk 8]
```diff
@@ -0,0 +1,16 @@
+from aleph_alpha_client.aleph_alpha_client import AlephAlphaClient
+from aleph_alpha_client.detokenization import DetokenizationRequest
+
+from tests.common import client, model
+
+
+def test_detokenize(client: AlephAlphaClient, model: str):
+    response = client.detokenize(model, request=DetokenizationRequest([4711]))
+
+    assert response.result is not None
```
[the remark this answers is not in the document]
Comment (ahartel) [reply tests/test_detokenize.py:10]: I think that would be akin to testing the tokenizer. In other words: If anyone changes the tokenizer then we need to adopt the test. I think we should avoid that.

File: aleph_alpha_client/completion.py (hunk 6 above)
Comment (ahartel) [reply aleph_alpha_client/completion.py:100]: I think it used to be that way and that's why I did not question that. But I added a test for that anyway and it seems to work.

File: tests/test_complete.py [hunk 9]
```diff
@@ -0,0 +1,27 @@
...
+        request=CompletionRequest(
+            prompt="", maximum_tokens=7, tokens=False, log_probs=0
+        ),
+    )
+
+    assert len(response.completions) == 1
+    assert response.model_version is not None
+
+
+def test_complete_with_explicit_parameters(client: AlephAlphaClient, model: str):
```
[the remark this answers is not in the document]
Comment (ahartel) [reply tests/test_complete.py:20]: donw

**Pull request merged**